core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java

```diff
@@ -0,0 +1,255 @@
+            }
+            // Set the parameters of the LISTAGG command within the state so that
+            // they can be used within the `output` function
+            state.setSeparator(separator);
+            state.setOverflowError(overflowError);
+            state.setOverflowFiller(overflowFiller);
+            state.setShowOverflowEntryCount(showOverflowEntryCount);
+        }
+        if (!value.isNull(position)) {
+            state.add(value, position);
+        }
+    }
+
+    public static void combine(Type type, ListaggAggregationState state, ListaggAggregationState otherState)
+    {
+        Slice previousSeparator = state.getSeparator();
+        if (previousSeparator == null) {
+            state.setSeparator(otherState.getSeparator());
+            state.setOverflowError(otherState.isOverflowError());
+            state.setOverflowFiller(otherState.getOverflowFiller());
+            state.setShowOverflowEntryCount(otherState.showOverflowEntryCount());
+        }
+
+        state.merge(otherState);
+    }
+
+    public static void output(Type type, ListaggAggregationState state, BlockBuilder out)
+    {
+        if (state.isEmpty()) {
+            out.appendNull();
+        }
+        else {
+            outputState(state, out, MAX_OUTPUT_LENGTH);
+        }
+    }
+
+    @VisibleForTesting
+    protected static void outputState(ListaggAggregationState state, BlockBuilder out, int maxOutputLength)
+    {
+        Slice separator = state.getSeparator();
+        int separatorLength = separator.length();
+        OutputContext outputContext = new OutputContext();
+        state.forEach((block, position) -> {
+            if (outputContext.outputLength > maxOutputLength) {
+                outputContext.outputLength += separatorLength + block.getSliceLength(position);
+            }
+            else {
+                if (outputContext.emittedEntryCount > 0) {
+                    if (outputContext.outputLength + separatorLength <= maxOutputLength) {
+                        // write the separator only if it fully fits within maxOutputLength
+                        out.writeBytes(separator, 0, separatorLength);
+                        outputContext.outputLength += separatorLength;
+                    }
+                    else {
+                        outputContext.outputLength += separatorLength;
+                        return;
+                    }
+                }
+
+                int entryLength = block.getSliceLength(position);
+                if (outputContext.outputLength + entryLength <= maxOutputLength) {
+                    // write the block only if it fully fits within maxOutputLength
+                    block.writeBytesTo(position, 0, entryLength, out);
+                    outputContext.emittedEntryCount++;
+                }
+                outputContext.outputLength += entryLength;
```

Comment:
Why is this outside of the `if` block? The length should be incremented only when the value is emitted.

Reply:
Same answert as https://github.com/trinodb/trino/pull/7992#discussion_r702148459